File: 1851.java
//kuannoodle的解
//思路与我不一样，他先将所有要查询的点query放到TreeMap里，然后用keySet得到TreeSet tset，是一个排了序的query点集合
//所有的interval按长度从小到大排序，然后对于每个区间intv，在tset中找到在这个区间中的点，这些点的最佳长度就更新了，更新以后从tset中移除掉
//这里完美的利用了TreeSet的方法 tset.subSet(start, end) 以及 集合的 tset.add(ele) tset.removeAll(tset2)
//这个方法的缺点是复杂度高，因为对interval和queries都进行了排序。
class Solution {
   public int[] minInterval(int[][] intervals, int[] queries) {

    final Map<Integer, List<Integer>> queryIndices = new HashMap<>();
    for (int i = 0; i != queries.length; i++) {
      final int v = queries[i];
      List<Integer> list = queryIndices.get(v);
      if (list == null)
        queryIndices.put(v, list = new ArrayList<>());
      list.add(i);
    }
    final TreeSet<Integer> tset = new TreeSet<>(queryIndices.keySet());

    int[] res = new int[queries.length];
    Arrays.fill(res, -1);

    Arrays.sort(intervals, (a, b) -> {
      return (a[1] - a[0]) - (b[1] - b[0]);
    });

    for (int[] intv : intervals) {
      final int intvlen = intv[1]+1-intv[0];
      List<Integer> removeem = new ArrayList<>();
      for (Integer queryval : tset.subSet(intv[0], intv[1] + 1)) {
        removeem.add(queryval);
        for (Integer qidx : queryIndices.get(queryval)) {
          res[qidx] = intvlen;
        }
      }
      tset.removeAll(removeem);
    }

    return res;

  }

}

File: 1851.py
#我的解
from heapq import heappop, heappush
from bisect import bisect_left
class Solution:
    def minInterval(self, intervals: List[List[int]], queries: List[int]) -> List[int]:
        for i in range(len(intervals)):
            if (intervals[i][0] > intervals[i][1]):
                intervals[i] = [intervals[i][1], intervals[i][0]]
        intervals.sort()
        h = []
        points = []
        for i, j in intervals:
            points.append(i)
            points.append(j)
        points = sorted(list(set(points)))
        cache = {}
        cache_p = {}
        interval_pos = 0
        for i, p in enumerate(points):
            while interval_pos < len(intervals) and (intervals[interval_pos][0] <= p):
                a,b = intervals[interval_pos]
                heappush(h, (b-a+1, a, b))
                interval_pos += 1
            while h[0][2] < p:
                heappop(h)
            cache_p[p] = h[0][0]
            while i+1 < len(points) and len(h) > 0 and h[0][2] < points[i+1]:
                heappop(h)
            if len(h) == 0:
                cache[p] = -1
            else:
                cache[p] = h[0][0]
                
        res = []
        for q in queries:
            if q in cache_p:
                res.append(cache_p[q])
            else:
                ind = bisect_left(points, q)
                if (ind == 0) or (ind == len(points)):
                    res.append(-1)
                else:
                    res.append(cache[points[ind-1]])
        return res
